NEXUS-5372: Alternative fix

cstamas opened the pull request:
Introduced new format, to replace the old broken
format, while keeping backward compat. In cases when
records are no longer than 255 bytes at least.

As in real life, records are longer than 255, this does not "salvage" the historical data (that is already written out corrupted), but at least does not break the build. Also, it uses proper Protobuf API to write out delimited records, is not rolling it's own to do so.

cstamas committed:
NEXUS-5372: Alternative fix

Introduced new format, to replace the old broken
format, while keeping backward compat. In cases when
records are no longer than 255 bytes at least.

ifedorenko commented:
-1

Lets remove V1 and V2 formats. V2 was broken and there is no point in providing support for it. V1 was before V2, so any historic timelines customers might have are broken.

Please provide a unit test for records longer than 255 bytes.

cstamas committed:
NEXUS-5372: Applied Igor's comments.

Changes:
* V1 and V2 support removed (they were busted anyway)
* All the UTs now operate with much higher records that before (intentionally
exceeding the problematic 255 bytes message length)
* UTs made stricter about expectations, as all the new "persist" data are actually
two "segments" with 53+47=100 records

cstamas commented:
Applied Igor's comments.

ifedorenko commented on nexus/plugins/nexus-timeline-plugin/src/main/java/org/sonatype/timeline/internal/DefaultTimelinePersistor.java:
```diff
@@ -106,14 +100,12 @@ protected synchronized void persist( final TimelineRecord... records )
         OutputStream out = null;
         try
         {
+            out = new FileOutputStream( getDataFile(), true );
```
I'd use BufferedOutputStream here.

ifedorenko commented on nexus/plugins/nexus-timeline-plugin/src/main/java/org/sonatype/timeline/internal/DefaultTimelinePersistor.java:
```diff
@@ -262,12 +254,12 @@ else if ( filePtr >= result.size() )
         try
         {
             in = new FileInputStream( file );
```
I'd use BufferedInputStream here

ifedorenko commented:
+1 with the comments above

cstamas committed:
NEXUS-5372: Using buffered IO

Applied Igor's comments, using Buffered IO
that makes much sense, as we are tossing
around "records" of few bytes.

nabcos commented:
+1
